lighthouse-core/lib/tracehouse/trace-processor.js

```diff
@@ -478,17 +480,67 @@ class TraceProcessor {
   }
 
   /**
-   * @param {{candidateEventName: string, invalidateEventName: string, events: LH.TraceEvent[], timeOriginEvt: LH.TraceEvent}} options
-   * @return {{lcp: LH.TraceEvent | undefined, invalidated: boolean}}
+   * @param {LH.TraceEvent} evt
+   * @return {evt is LCPEvent}
    */
-  static computeValidLCP(options) {
-    const {
-      candidateEventName,
-      invalidateEventName,
-      events,
-      timeOriginEvt,
-    } = options;
+  static isLCPEvent(evt) {
+    if (evt.name !== 'largestContentfulPaint::Invalidate' &&
+        evt.name !== 'largestContentfulPaint::Candidate') return false;
+    return !!(evt.args && evt.args.frame);
+  }
+
+  /**
+   * @param {LH.TraceEvent} evt
+   * @return {evt is LCPCandidateEvent}
+   */
+  static isLCPCandidateEvent(evt) {
+    if (evt.name !== 'largestContentfulPaint::Candidate') return false;
+    return !!(evt.args && evt.args.frame && evt.args.data && evt.args.data.size !== undefined);
+  }
+
+  /**
+   * @param {LH.TraceEvent[]} events
+   * @param {LH.TraceEvent} timeOriginEvent
+   * @return {LCPEvent | undefined}
+   */
+  static computeValidLCPAllFrames(events, timeOriginEvent) {
+    const lcpEvents = events.filter(this.isLCPEvent).reverse();
```

Comment:
Just noting that the reason the other fn manually iterates backwards instead of using `Array.reverse` is because `events` is huge-although I see you filter the events here, so it's all good.